### PR TITLE
InstSimplify: lookthru casts, binops in folding shuffles

### PR DESCRIPTION
**Pull request by artagnon**
foldIdentityShuffles() currently suffers from the deficiency that it cannot look through casts or binary operators, and can only fold sequences of shuffles without any intervening instructions. Fix this deficiency by looking through chains of casts that preserve the number of elements in the vector, and chains of binary operators with identical operands. The way we do this is by saving the last instruction in the reduction chain as the RootVec, and rewriting the operands of the first instruction in the chain.

This patch also has a requirement on increasing the MaxRecursionLimit, and the AMDGPU and LoopVectorize test changes are due to that.

Future work could simplify the case of binary operands with different operands, and the multi-use case.

Alive2 proofs follow.

  @fold_lookthrough_cast:
    https://alive2.llvm.org/ce/z/YFKQGX
  @fold_lookthrough_binop_same_operands:
    https://alive2.llvm.org/ce/z/gf3WgB
  @fold_lookthrough_cast_chain:
    https://alive2.llvm.org/ce/z/MT9ApH
  @fold_lookthrough_binop_chain:
    https://alive2.llvm.org/ce/z/WV8HRn
  @fold_lookthrough_cast_binop_chain:
    https://alive2.llvm.org/ce/z/CXj-hh

-- 8<--
Based on #92407.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/select-constant-cttz.ll
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/select-constant-cttz.ll:9]: This looks like the point of the test was defeated, so should do something to keep the codegen similar

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:5524]: C++17 structured binding

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:5394]: dyn_cast instead of isa+cast

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:5352]: Braces

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (artagnon) [reply llvm/lib/Analysis/InstructionSimplify.cpp:5524]: Doesn't improve the code in any meaningful way, as we need to update `RootVec` and `ReplaceInRootVec`, which are declared outside the loop.

File: llvm/test/CodeGen/AMDGPU/select-constant-cttz.ll
Comment (artagnon) [reply llvm/test/CodeGen/AMDGPU/select-constant-cttz.ll:9]: Not sure what we can do about this. Want to take a stab at making sure the test doesn't get reduced? Sorry, I'm not familiar with AMDGPU.

**Comment by artagnon** [comment]
Why is the CI failing? All tests pass, so is it some timing issue?

File: llvm/test/Transforms/InstSimplify/shufflevector.ll
Comment (davemgreen) [inline llvm/test/Transforms/InstSimplify/shufflevector.ll:343]: Would this need to remove the shuffle too?

**Review by davemgreen — COMMENTED** [review]
Hi - I recently added something very similar in VectorCombine under #88693. I deliberately added it to vector combine as I thought trying the combine many times through the pipeline was best avoided. I didn't think we had things like that in instsimplify.

I have some patches to extend the current functionality, and it should be able to handle this case if I understand it correctly. I have been holding off adding additions until the comments from https://github.com/llvm/llvm-project/commit/a4d10266d20bfe5930dfed77e17832af341ed66e can be worked out. I can try and put up the zext/sext/trunc patch I have, it should already handle intrinsics and binops/other unary ops.

> Why is the CI failing? All tests pass, so is it some timing issue?

That looks like ccache failed - it sometimes does that.

File: llvm/test/Transforms/InstSimplify/shufflevector.ll
Comment (davemgreen) [reply llvm/test/Transforms/InstSimplify/shufflevector.ll:343]: Oh - I see it is now unused.

**Comment by artagnon** [comment]
I had a look at #88693, and the approach seems a lot more complicated than the one in this patch. I think there's a clear separation of concerns though: InstCombine (and by extension, InstSimplify) is equipped to handle only simple cases, performing trivial removal and rewriting, where the transform is obviously profitable, while VectorCombine is geared toward more complicated transforms possibly involving intrinsics, rewriting the entire tree of instructions, and using a CostModel.

For instance, the case of binary operators with unequal operands and intrinsics can be handled in VectorCombine, but for the simple case of just looking through some intermediate instructions between canceling shuffles, and eliminating the shuffles can be done in InstSimplify.

Besides, there was already code to fold canceling shuffles in InstSimplify, and there are several existing tests in InstSimplify/InstCombine that exercise this. I merely extended it a little bit.

What do you think?

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (dtcxzyw) [inline llvm/lib/Analysis/InstructionSimplify.cpp:51]:
```suggestion
static constexpr unsigned RecursionLimit = 5;
```

@nikic Can you test the compile-time impact?

TBH I don't like increasing the recursion limit for all root instructions.
If necessary, it would be better to only increase the limit for `shufflevector`:
```
/// Given operands for a ShuffleVectorInst, fold the result or return null.
Value *llvm::simplifyShuffleVectorInst(Value *Op0, Value *Op1,
                                       ArrayRef<int> Mask, Type *RetTy,
                                       const SimplifyQuery &Q) {
  static constexpr ShuffleRecursionLimit = 5;
  return ::simplifyShuffleVectorInst(Op0, Op1, Mask, RetTy, Q, ShuffleRecursionLimit);
}
```

Comment (arsenm) [reply llvm/lib/Analysis/InstructionSimplify.cpp:51]: I'm surprised there's a threshold here that isn't 6

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:5413]: Checking for users in InstSimplify feels weird, can you check for hasOneUse on the input instead/

File: llvm/test/CodeGen/AMDGPU/select-constant-cttz.ll
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/select-constant-cttz.ll:9]: The AMDGPU part doesn't matter, just make sure whatever simplify now happens, still doesn't happen. It's super frustruating that InstSimplify ends up breaking codegen tests like this. 

Is this just from the recursive depth increase?

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (nikic) [reply llvm/lib/Analysis/InstructionSimplify.cpp:51]: Compile-time: http://llvm-compile-time-tracker.com/compare.php?from=f6527774569790b5a5236f6e84f3f839ce6c2fff&to=4363c34b550a0908edf4ee89eda2fee72a096ba9&stat=instructions:u

**Review by nikic — CHANGES_REQUESTED** [review]
You cannot RAUW inside InstSimplify. InstSimplify is an analysis, it cannot change IR in any way.

**Comment by artagnon** [comment]
> You cannot RAUW inside InstSimplify. InstSimplify is an analysis, it cannot change IR in any way.

Oh no. Looks like I have to clone the entire chain of instructions, make the replacement in the new head instruction, and return the new tail? Is InstSimplify poorly-suited for this kind of change, and should I just leave it up to VectorCombine?

File: llvm/test/CodeGen/AMDGPU/select-constant-cttz.ll
Comment (artagnon) [reply llvm/test/CodeGen/AMDGPU/select-constant-cttz.ll:9]: Yeah, it's due the increase in recursion depth, but we can limit it to the shuffle, as @dtcxzyw suggested.

**Comment by arsenm** [comment]
> > You cannot RAUW inside InstSimplify. InstSimplify is an analysis, it cannot change IR in any way.
> 
> Oh no. Looks like I have to clone the entire chain of instructions, make the replacement in the new head instruction, and return the new tail? Is InstSimplify poorly-suited for this kind of change, and should I just leave it up to VectorCombine?

You shouldn't be creating new instructions in instsimplify. It's for returning operands or existing instructions

**Comment by artagnon** [comment]
> You shouldn't be creating new instructions in instsimplify. It's for returning operands or existing instructions

Okay, abandoning this PR, as the task is impossible to do within instsimplify. I'll move the tests to VectorCombine, and work with @davemgreen on his code.

**Comment by nikic** [comment]
@artagnon See https://llvm.org/docs/InstCombineContributorGuide.html#pick-the-correct-optimization-pass for basic guidance on which pass to put things. Putting this in InstCombine may be fine -- but if VectorCombine support is already on the way, no point in doing that.

**Pull request closed without merging**
